### PR TITLE
Add drop=True argument to isel, sel and squeeze

### LINKED ISSUE #242: Add a "drop" option to squeeze
If True, squeezed dimensions should be dropped from the resulting object (instead of being retained as scalar)

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #242

This is useful for getting rid of extraneous scalar variables that arise from
indexing, and in particular will resolve an issue for optional indexes:
https://github.com/pydata/xarray/pull/1017#issuecomment-260777664

### CODE REVIEW

File: xarray/test/test_dataset.py [hunk 1]
```diff
@@ -1750,6 +1770,16 @@ def get_args(v):
...
+    def test_squeeze_drop(self):
+        data = Dataset({'foo': ('x', [1])}, {'x': [0]})
+        expected = Dataset({'foo': 1})
+        selected = data.squeeze(drop=True)
+        self.assertDatasetIdentical(expected, selected)
+
+        expected = Dataset({'foo': 1}, {'x': 0})
+        selected = data.squeeze(drop=False)
+        self.assertDatasetIdentical(expected, selected)
+
```
Comment (jhamman) [inline xarray/test/test_dataset.py]: Can you add a few more tests to this section:

- squeeze empty array
- squeeze 2d array with a single element

Comment (shoyer) [inline xarray/test/test_dataset.py]: done

**Comment by shoyer** [comment]
Going to merge this shortly unless anyone has objections

**Comment by pwolfram** [comment]
@shoyer, I think this is really great and appreciate you doing this awesome work.  I obviously have a vested interest in the `drop=True` capability because we have found great use for it via `where`. My vote is go for it!

**Comment by shoyer** [comment]
>  I obviously have a vested interest in the drop=True capability because we have found great use for it via where.

Well, to be clear `drop=True` works pretty differently on these methods than in `where`, so this might be confusing.

For `isel`, `sel` and `squeeze`, it drops scalar coordinates variables that would arise from indexing, whereas for `where` it drops some coordinate values. That said, the `drop` method already handles both cases.

**Pull request merged**